doc/manual/src/architecture/file-system-object.md

```diff
@@ -0,0 +1,63 @@
+# File System Object
+
+Nix uses a simplified model of the file system, which consists of file system objects.
+Every file system object is one of the following:
+
+ - File
+
+   - A possibly empty sequence of bytes for contents
+   - A single boolean representing the [executable](https://en.m.wikipedia.org/wiki/File-system_permissions#Permissions) permission
+
+ - Directory
+
+   Mapping of names to child file system objects
+
+ - [Symbolic link](https://en.m.wikipedia.org/wiki/Symbolic_link)
+
+   A file system path that may point anywhere
+
+File systems objects and their children form a tree.
+A bare file or symlink can be a root file system object.
+
+## Examples of file system objects
+
+A plain file:
+
+```
+. "Hello World", executable: false
+```
+
+An executable file:
+```
+. "#!/bin/bash\necho Hello World", executable: true
+```
```

Comment:
````suggestion
```
"Hello World", executable: false
```

An executable file:
```
"#!/bin/bash\necho Hello World", executable: true
```
````

`.` is part of the directory tree syntax. We don't have a directory, so we don't have a directory tree either.

Reply:
That `.` is not actually printed by `tree`, we introduced it to signify the root simply to have some indication that it's not just some random string but relates to the other examples.

Reply:
[`tree`](https://linux.die.net/man/1/tree) prints these trees. I know, imaginary syntax is not great, but we're already imagining some representation here. We could also drop the examples entirely...

Reply:
I did like the examples, but maybe you're right. Better not to make up any syntax at all. I could make a joke about ATerm now, but instead I'll make one about UML. Actually it doesn't fit in the margin though.

We could do it in Nix with a `_type` attribute to represent the "sum type", but it's not nice to suggest that such values would be a thing, and Nix strings don't represent all bytes.

Let's discuss again when we write the Nix specification, and make any decision for now.

Reply:
We could also just explain the imaginary syntax.

Reply:
It's already too long for a glossary item. If we're going to write more, we should create a proper page for it.

Reply:
It is a proper page already, the glossary item is just for discoverability.